Add new mlxconfig params to mellanox_fw_update.yaml

wmousa opened the pull request:
(ESWITCH_HAIRPIN_TOT_BUFFER_SIZE, ESWITCH_HAIRPIN_DESCRIPTORS)

wmousa committed:
Add new mlxconfig params to mellanox_fw_update.yaml
(ESWITCH_HAIRPIN_TOT_BUFFER_SIZE, ESWITCH_HAIRPIN_DESCRIPTORS)

wmousa commented:
@adrianchiris @moshe010 please review this

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -40,6 +40,21 @@ parameters:
+                 If you need to set configuration for specific indexes you can pass the
+                 value as below for index 2 to be 17 and index 3 to be 16
+                 Example, {"2": "17", "3": "16"}
+                 Make sure to choose only the PCI ends with 0
```
>  Make sure to choose only the PCI ends with 0

if its a global/per-host parameter it shouldnt matter right ?
also if you run on more than one PF, we are not appying the configuration twice correct?

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -40,6 +40,21 @@ parameters:
     description: 'Priority tag required'
+  ESWITCH_HAIRPIN_TOT_BUFFER_SIZE:
+    type: json
+    default: {"*": "17"}
```
better than a list :)

moshe010 commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -230,7 +263,30 @@ resources:
                             continue
 
-                        if current_mlx_config.get(key) and value.lower(
+                        if key in self.arrayed_mlnx_config.keys():
```
remove the keys()
if key in self.arrayed_mlnx_config

moshe010 commented on templates/firstboot/mlnx_fw_update.py:
```diff
@@ -16,25 +16,28 @@
 FW_VERSION_REGEX = r'FW Version:\s*\t*(?P<fw_ver>\d+\.\d+\.\d+)'
 RUNNING_FW_VERSION_REGEX = r'FW Version\(Running\):\s*\t*(?P<fw_ver>\d+\.\d+\.\d+)'
 PSID_REGEX = r'PSID:\s*\t*(?P<psid>\w+)'
+ARRAY_VALUE_REGEX = r'Array\[(?P<first_index>\d+)\.\.(?P<last_index>\d+)\]'
```
not for this PR. but in the future we should create the Yaml file automatically from this python script using jinja to render from template. Then you should maintain the script in the yaml file. Again this is not for now, but for future enchantment

adrianchiris replied:
tbh the script is getting pretty large. if it keeps growing we would need to explore alternatives to just putting it flat in the yaml

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -40,6 +40,21 @@ parameters:
+  ESWITCH_HAIRPIN_TOT_BUFFER_SIZE:
+    type: json
+    default: {"*": "17"}
+    description: If "*" is on of the json keys, then its value will set to all indexes
```
i would reword
`If a single key of "*" is provided, then its value will be set to all indexes`

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -40,6 +40,21 @@ parameters:
+  ESWITCH_HAIRPIN_DESCRIPTORS:
+    type: json
+    default: {"*": "11"}
+    description: If "*" is on of the json keys, then its value will set to all indexes
```
i would reword
`If a single key of "*" is provided, then its value will be set to all indexes`

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -193,16 +212,30 @@ resources:
 
-                def get_device_conf_dict(self):
-                    """ Get device Configurations
+                def get_device_conf_dict(self, param_name=None):
```
The flow is a bit convoluted, having `get_device_conf_dict` call `_mstconfig_parse_data` which calls back `get_device_conf_dict`

you could modify get_device_conf_dict to perform a 2 passess on the output. 
1 the normal mlxconfig query
2. expand the array configuration to their elements (you can do that recursively if you prefer using `get_device_conf_dict`)

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -181,6 +199,7 @@ resources:
                 def __init__(self, pci_dev):
                     self.pci_dev = pci_dev
                     self._tool_confs = None
+                    self.arrayed_mlnx_config = {}
```
maybe add a comment explaining what this attribute stores/represents ? or better yet just define a class for it (which has first and last index attributes)

also perhaps a better name would be:  `array_config_info`  ?

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -193,16 +212,30 @@ resources:
+                        if array_value:
+                            first_index = array_value.group('first_index')
+                            last_index =  array_value.group('last_index')
+                            self.arrayed_mlnx_config[d[0]] = [first_index, last_index]
```
i prefer not to hide filling up the self.arrayed_mlnx_config in _mst_config_parse_data

can we have a separate method that gets the conf_dict and fills up that information ?

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -230,7 +263,30 @@ resources:
 
-                        if current_mlx_config.get(key) and value.lower(
+                        if key in self.arrayed_mlnx_config.keys():
+                            if value.get('*'):
```
id prefer if you implement the logic to expand the user provided "conf_dict" in a separate method.
also  need to update the docstring for set_config.

adrianchiris replied:
need to validate that if "*" is provided, no other entries are in the dict.

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -230,7 +263,30 @@ resources:
+                                 for key_index, param_value in value.items():
+                                     param_key = "%s[%s]" % (key, key_index)
+                                     if current_mlx_config.get(param_key) and \
+                                             str(param_value).lower() not in \
```
is there a reason you are using `in` and not `==` ?
e.g  "1" in "100" is true and would then lead to the configuration to be skipped

adrianchiris commented on templates/firstboot/mellanox_fw_update.yaml:
```diff
@@ -230,7 +263,30 @@ resources:
+                                                                         param_value))
+
+                        elif current_mlx_config.get(key) and value.lower(
                         ) not in current_mlx_config.get(key).lower():
```
same issue is here as well i think

adrianchiris committed:
Update Code according to comments

1. define a new array parameter metadata class
2. refactor the logic and avoid recursion

adrianchiris commented:
@moshe010 @wmousa

moshe010 commented on templates/firstboot/mlnx_fw_update.py:
```diff
@@ -157,9 +180,77 @@ def param_supp_by_config_tool(self, param_name):
         return param_name in self._tool_confs
 
+    def _build_config_param_metadata_map(self, conf_dict):
+        self.mlnx_config_array_param_metadata = dict()
```
can you make all the code use {} instead of dict() just be consistent

adrianchiris replied:
sure

moshe010 reviewed (commented):
the change is very complex ;(. you probably tested only the script, right?
I guess Waleed will need to test  full deployment ...

adrianchiris committed:
Some minor enhancements

1. Add log message when user provided an index out of range for array
parameter
2. Fail with runtime error if user provided mixed input of "*" and
specific index for an array parameter as they are mutualy exclusive

wmousa committed:
Add some fixes to the code

adrianchiris reviewed (commented):
LGTM
